synapse/storage/databases/main/purge_events.py

```diff
@@ -283,6 +281,18 @@ def _purge_history_txn(
         # so make sure to keep this actually last.
         txn.execute("DROP TABLE events_to_purge")
 
+        for event_id, should_delete in event_rows:
+            self._invalidate_cache_and_stream(
+                txn, self._get_state_group_for_event, (event_id,)
+            )
+
+            # FIXME: this is racy - what if have_seen_event gets called between the
+            #    transaction completing and the invalidation running?
```

Comment:
I'm not sure there is much difference between that and calling `have_seen_event` just before we delete the event?

Reply:
true. I've updated the comment accordingly.